perun/collect/trace/locks.py

```diff
@@ -0,0 +1,231 @@
+    :ivar str res: the name of the Res index corresponding to the type of the resource
+    :ivar int pid: the PID of the locking process
+    :ivar str locks_dir: the path to the .perun directory where lock files are stored
+    :ivar str file: the full path of the resulting lock file
+    """
+    def __init__(self, resource_type, resource_name, pid, locks_dir):
+        """ Construct lock object
+
+        :param LockType resource_type: the type of the resource to lock
+        :param str resource_name: the name of the lock (e.g. the name of the profiled binary file)
+        :param int pid: the PID of the locking process
+        :param str locks_dir: the path to the .perun directory where lock files are stored
+        """
+        self.name = resource_name
+        self.type = resource_type
+        self.res = resource_type.value['res']
+        self.pid = pid
+        self.locks_dir = locks_dir
+        self.file = os.path.join(
+            locks_dir, '{}:{}{}'.format(self.name, self.pid, self.type.value['suffix'])
+        )
+
+    @classmethod
+    def fromfile(cls, lock_file):
+        """ Construct ResourceLock object from a lock file
+
+        :param str lock_file: the path of the lock file
+        :return ResourceLock or None: the lock object or None if the file does not represent a lock
+        """
+        try:
+            # Get the resource name and pid
+            name, rest = os.path.basename(lock_file).rsplit(':', maxsplit=1)
+            pid = int(rest[:-LOCK_SUFFIX_LEN])
+            # Transform the suffix into a LockResourceType
+            resource_type = LockType.suffix_to_type(rest[-LOCK_SUFFIX_LEN:])
+            if resource_type is None:
+                # Invalid resource type = invalid resource lock
+                return None
+            return cls(resource_type, name, pid, os.path.dirname(lock_file))
+        except ValueError:
+            return None
+
+    def lock(self, res):
+        """ Actually locks the resource represented by the lock object.
+
+        :param Res res: the Res object responsible for keeping track of resources
+        """
+        WD.debug("Attempting to lock a resource '{}' with pid '{}'".format(self.name, self.pid))
+
+        # Lock the resource first
+        temp.touch_temp_dir(self.locks_dir)
+        res[self.res] = self
+        temp.touch_temp_file(self.file, protect=True)
+        # Check that no other currently running profiling process has locked the same resource
+        # The check should be done again later since data race might have happened
+        self.check_validity()
+
+        WD.debug("Resource locked: '{}'".format(self.file))
+
+    @staticmethod
+    def unlock(lock, res):
+        """ Unlocks the resource represented by the lock object.
+
+        :param ResourceLock lock: the lock object
+        :param Res res: the Res object responsible for keeping track of resources
+        """
+        if lock is not None:
```

Comment:
Suggetsion: not lock?

Reply:
The lock parameter can either be a ResourceLock object or None, and for such values, a `is not None` comparison is recommended by the PEP.